src/librustc_trait_selection/opaque_types.rs

```diff
@@ -716,24 +716,34 @@ where
             ty::Closure(_, ref substs) => {
                 // Skip lifetime parameters of the enclosing item(s)
 
-                for upvar_ty in substs.as_closure().upvar_tys() {
-                    upvar_ty.visit_with(self);
-                }
+                if !substs.as_closure().is_valid() {
```

Comment:
check infcx here? 

you can have take a `infcx` to ConstrainOpaqueTypeRegionVisitor. And since the place where u construct the visitor is while implementing the inference context you can just pass self.

Reply:
The reason for this is, the trait selection code runs before and after typechk and is_valid doesn't gauranteee the correct result until after typechk is complete.